### PR TITLE
Add proper contexts with timeouts for etcd operations

### LINKED ISSUE #7096: etcd endpoints is not running, starting minio server in FS mode will hang
I can confirm this. If the etcd endpoints are specified in ENV and etcd is not running, starting minio server in FS mode will hang with no output and cannot be terminated by ^c.

Steps to reproduce:
export MINIO_ETCD_ENDPOINTS=http://1.1.1.1:2379

minio server --address :80 /tmp
<no output>
^C^C^C^C

_Originally posted by @dustinaevans in https://github.com/minio/minio/issues/6554#issuecomment-454804910_

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Add proper contexts with timeouts for etcd operations

## Motivation and Context
This fixes an issue of perceived hang when incorrect
unreachable URLs are specified in MINIO_ETCD_ENDPOINTS
variable.

Fixes #7096

## Regression
No

## How Has This Been Tested?
Using the example posted in issue #7096 

```
~ MINIO_ETCD_ENDPOINTS=http://1.1.1.1:2379 minio server ~/test
... wait for 30 secs ...
ERROR Unable to initialize config system: etcd setup is unreachable, please check your endpoints [http://1.1.1.1:2379].
```

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :x: [more...](#b2bbea8d-0fba-45f9-ab54-ae37f3d0e9e2) |
<a name="b2bbea8d-0fba-45f9-ab54-ae37f3d0e9e2"></a>
### 7097-b50fbd0/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32121
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d24da9241b09:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 58 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 4 seconds
(4/13) Running aws-sdk-php tests ... done in 46 seconds
(5/13) Running aws-sdk-ruby tests ... done in 12 seconds
(6/13) Running mc tests ... done in 1 minutes and 14 seconds
(7/13) Running minio-dotnet tests ... FAILED in 5 minutes and 53 seconds

Unhandled Exception: System.AggregateException: One or more errors occurred. (One or more errors occurred. (Minio API responded with message=We encountered an internal error, please try again.)) ---> System.AggregateException: One or more errors occurred. (Minio API responded with message=We encountered an internal error, please try again.) ---> Minio.Exceptions.MinioException: Minio API responded with message=We encountered an internal error, please try again.
   at Minio.MinioClient.ParseError(IRestResponse response) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 363
   at Minio.MinioClient.HandleIfErrorResponse(IRestResponse response, IEnumerable`1 handlers, DateTime startTime) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 505
   at Minio.MinioClient.ExecuteTaskAsync(IEnumerable`1 errorHandlers, IRestRequest request, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 351
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, String uploadId, Int32 partNumber, Byte[] data, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 488
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, Stream data, Int64 size, String contentType, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 234
   at Minio.Functional.Tests.FunctionalTest.PutObject_Task(MinioClient minio, String bucketName, String objectName, String fileName, String contentType, Int64 size, Dictionary`2 metaData, MemoryStream mstream) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 740
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.RemoveObjects_Test2(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 1738
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.Main(String[] args) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 218

Executed 6 out of 13 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

File: cmd/config-common.go [hunk 1]
```diff
@@ -60,8 +61,15 @@ func deleteConfig(ctx context.Context, objAPI ObjectLayer, configFile string) er
 }
 
 func saveConfigEtcd(ctx context.Context, client *etcd.Client, configFile string, data []byte) error {
-	_, err := client.Put(ctx, configFile, string(data))
-	return err
+	timeoutCtx, cancel := context.WithTimeout(ctx, defaultContextTimeout)
+	_, err := client.Put(timeoutCtx, configFile, string(data))
+	defer cancel()
```
Comment (poornas) [inline cmd/config-common.go:65]: defer cancel() should be done before etcd operation

File: cmd/config-common.go [hunk 2]
```diff
@@ -75,9 +83,14 @@ func saveConfig(ctx context.Context, objAPI ObjectLayer, configFile string, data
 }
 
 func readConfigEtcd(ctx context.Context, client *etcd.Client, configFile string) ([]byte, error) {
-	resp, err := client.Get(ctx, configFile)
+	timeoutCtx, cancel := context.WithTimeout(ctx, defaultContextTimeout)
+	resp, err := client.Get(timeoutCtx, configFile)
+	defer cancel()
```
Comment (poornas) [inline cmd/config-common.go:87]: switch lines 87/88

File: cmd/config-common.go [hunk 3]
```diff
@@ -104,9 +119,14 @@ func watchConfig(objAPI ObjectLayer, configFile string, loadCfgFn func(ObjectLay
 }
 
 func checkConfigEtcd(ctx context.Context, client *etcd.Client, configFile string) error {
-	resp, err := globalEtcdClient.Get(ctx, configFile)
+	timeoutCtx, cancel := context.WithTimeout(ctx, defaultContextTimeout)
+	resp, err := client.Get(timeoutCtx, configFile)
+	defer cancel()
```
Comment (poornas) [inline cmd/config-common.go:123]: switch lines 123/124

File: pkg/quick/encoding.go [hunk 4]
```diff
@@ -138,18 +138,26 @@ func saveFileConfigEtcd(filename string, clnt *etcd.Client, v interface{}) error
...
+		return fmt.Errorf("unexpected error %s returned by etcd setup, please check your endpoints %s", err, clnt.Endpoints())
+	}
+	return nil
 }
 
 func loadFileConfigEtcd(filename string, clnt *etcd.Client, v interface{}) error {
-	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
 	resp, err := clnt.Get(ctx, filename)
 	defer cancel()
```
Comment (poornas) [inline pkg/quick/encoding.go:154]: switch lines 154/155

**Review by poornas — CHANGES_REQUESTED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by vadmeste — APPROVED** [review]
LGTM & tested with two Minio instances and etcd server

**Review by poornas — APPROVED** [review]
LGTM - tested g/w mode with etcd server down

**Comment by markmarijnissen** [comment]
How should minio handle unavailable etcd servers? It seems that this silently fails and propagates errors to etcd-dependent functions. 

**My use case**
I run Minio 3.3.10 in gateway mode with an etcd endpoint on kubernetes.

When k8s restarts some pods, both minio and etcd come back online as they should. However, for some reason, minio fails to connect to etcd. After several attempts, it seems to 'give up'.

When I try to use crendentials or admin commands, it fails with more or less obscure error messages, like "the api is not available" or "the client and server version do not match".

**Pull request merged**